pv_site_production/app.py

```diff
@@ -2,14 +2,126 @@
+        self._meta = {
+            # TODO Standardize the column names.
+            # Making sure the pv_id (ss_id) is parsed as an integer.
+            int(row["ss_id"]): {key: row[key] for key in META_FILE_KEYS}
+            for _, row in pd.read_csv(metadata_path).iterrows()
+        }
+
+    def get(
+        self,
+        pv_ids: list[PvId] | PvId,
+        start_ts: Timestamp | None = None,
+        end_ts: Timestamp | None = None,
+    ) -> xr.Dataset:
+        session = self._db_connection.get_session()
+
+        # TODO The fact that we have to check for two types tells me something does not get checked
+        # properly somewhere!
+        if isinstance(pv_ids, (PvId, np.integer)):
+            was_1d = True
+            pv_ids = [pv_ids]
+        else:
+            was_1d = False
+
+        pv_yields = get_pv_yield(
+            session=session,
+            pv_systems_ids=pv_ids,
+            start_utc=start_ts,
+            end_utc=end_ts,
+            # TODO What does this option mean?
+            correct_data=None,
+            providers=[SHEFFIELD],
+        )
+
+        records = [(PVYield.from_orm(pv_yield)).__dict__ for pv_yield in pv_yields]
+
+        # Build a pandas dataframe of pv_id, timestamp and power. This makes it easy to convert to
+        # an xarray.
+        df = pd.DataFrame.from_records(
+            {
+                "pv_id": r["pv_system"].pv_system_id,
+                # We remove the timezone information since otherwise the timestamp index gets
+                # converted to an "object" index later. In any case we should have everything in
+                # UTC.
+                "ts": r["datetime_utc"].replace(tzinfo=None),
+                "power": r["solar_generation_kw"],
+            }
+            for r in records
+        )
+        # Convert it to an xarray.
+        df = df.set_index(["pv_id", "ts"])
+        da = xr.Dataset.from_dataframe(df)
+
+        # Add the metadata associated with the PV systems.
+        # Some come from the database, and some from a separate metadata file.
+        # TODO The info from the metadata file (tilt, orientation and "factor")
+        # probably belong to the database!
+        meta_from_db = {
+            r["pv_system"].pv_system_id: {
+                key: getattr(r["pv_system"], key) for key in META_DB_KEYS
+            }
+            for r in records
+        }
+
+        pv_ids = [int(x) for x in da.coords["pv_id"].values]
+
+        # Merge both the meta data from the DB and from the file.
+        meta = {pv_id: self._meta[pv_id] | meta_from_db[pv_id] for pv_id in pv_ids}
+
+        # Add the metadata as coordinates to the pv_ids in the xr.Dataset.
+        da = da.assign_coords(
+            {
+                key: ("pv_id", [meta[pv_id][key] for pv_id in pv_ids])
+                for key in META_FILE_KEYS + META_DB_KEYS
+            }
+        )
+
+        if was_1d:
+            da = da.isel(pv_id=0)
+
```

Comment:
I think this is all fine, until it dosnt work, becasue data is delayed e.t.c. Thats where using the `ocf_datapipes` will be handy. 

Of course you still perhaps need to rename variables how you wish. 

Perhaps I could add a filter for `pv_ids` into the `ocf_datapipes` module so you have that functionality